### PR TITLE
Gemmlowp deps build

### PR DESCRIPTION
**Pull request by harouwu**
This is based on PR https://github.com/pytorch/pytorch/pull/12947 However, CI takes too long to test so I put it here

### CODE REVIEW

File: cmake/Dependencies.cmake [hunk 1]
```diff
@@ -108,6 +108,18 @@ if(USE_NNPACK)
   endif()
 endif()
 
+# ---[ QNNPACK, use gemmlowp
+if(USE_GEMMLOWP)
```
Comment (Maratyszcza) [inline cmake/Dependencies.cmake]: `USE_INT8_OPS` would be a better name as QNNPACK itself doesn't depend on gemmlowp

File: cmake/External/gemmlowp.cmake [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+##############################################################################
+# (2) Anything but x86, x86-64, ARM, ARM64 - unsupported
+##############################################################################
+if(CMAKE_SYSTEM_PROCESSOR)
+  if(NOT CMAKE_SYSTEM_PROCESSOR MATCHES "^(i686|x86_64|armv5te|armv7-a|armv7l|aarch64)$")
+    message(WARNING "gemmlowp is not supported on ${CMAKE_SYSTEM_PROCESSOR} processors. "
+      "The only supported architectures are x86, x86-64, ARM, and ARM64. "
+      "Turn this warning off by USE_GEMMLOWP=OFF.")
+    set(USE_GEMLOWP OFF)
```
Comment (Maratyszcza) [inline cmake/External/gemmlowp.cmake]: Typo, should be `USE_GEMMLOWP`

**Comment by harouwu** [comment]
Fixed comments and fixed gemmlowp include dir issue

File: CMakeLists.txt [hunk 3]
```diff
@@ -97,6 +97,7 @@ option(USE_NCCL "Use NCCL" ON)
 option(USE_SYSTEM_NCCL "Use system-wide NCCL" OFF)
 option(USE_NNAPI "Use NNAPI" OFF)
 option(USE_NNPACK "Use NNPACK" ON)
+option(USE_GEMMLOWP "Use gemmlwop" ON)
```
Comment (Maratyszcza) [inline CMakeLists.txt]: typo!

File: cmake/Dependencies.cmake [hunk 4]
```diff
@@ -108,6 +108,15 @@ if(USE_NNPACK)
   endif()
 endif()
 
+# ---[ Caffe2 Int8 Ops, use gemmlowp
+if(USE_GEMMLOWP)
+  include(${CMAKE_CURRENT_LIST_DIR}/External/gemmlowp.cmake)
+  if(NOT GEMMLOWP_FOUND)
+    message(WARNING "Not compiling with NNPACK. Suppress this warning with -DUSE_GEMMLOWP=OFF")
```
Comment (Maratyszcza) [inline cmake/Dependencies.cmake]: Should be "Not compiling with GEMMLOWP"

File: cmake/External/gemmlowp.cmake [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    return()
+  endif()
+endif()
+
+##############################################################################
+# (3) Android, iOS, Linux, macOS - supported
+##############################################################################
+
+if (ANDROID OR IOS OR ${CMAKE_SYSTEM_NAME} STREQUAL "Linux" OR ${CMAKE_SYSTEM_NAME} STREQUAL "Darwin")
+  message(STATUS "Brace yourself, we are building gemmlowp")
```
Comment (Maratyszcza) [inline cmake/External/gemmlowp.cmake]: We are not actually building gemmlowp, so better remove this message :)

File: setup.py [hunk 6]
```diff
@@ -177,6 +177,7 @@ def hotpatch_var(var, prefix='USE_'):
 from tools.setup_helpers.mkldnn import (USE_MKLDNN, MKLDNN_LIBRARY,
                                         MKLDNN_LIB_DIR, MKLDNN_INCLUDE_DIR)
 from tools.setup_helpers.nnpack import USE_NNPACK
+from tools.setup_helpers.qnnpack import USE_GEMMLOWP
```
Comment (Maratyszcza) [inline setup.py]: This is confusing. Lets use `USE_QNNPACK` variable for Int8

File: setup.py [hunk 7]
```diff
@@ -383,6 +384,8 @@ def build_libs(libs):
         build_libs_cmd += ['--use-rocm']
     if USE_NNPACK:
         build_libs_cmd += ['--use-nnpack']
+    if USE_GEMMLOWP:
```
Comment (Maratyszcza) [inline setup.py]: same here

**Comment by Maratyszcza** [comment]
I suggest you just unconditionally add gemmlowp include directory. We'll handle supported/unsupported platforms in QNNPACK dependency.

**Review by Maratyszcza — APPROVED** [review]
LG, assuming comments are addressed

**Comment by harouwu** [comment]
Comments addressed :)

**Pull request closed without merging**